core/dbt/parser/manifest.py

```diff
@@ -243,7 +243,11 @@ def __init__(
         self.root_project: RuntimeConfig = root_project
         self.all_projects: Mapping[str, Project] = all_projects
         self.file_diff = file_diff
-        self.publications = publications
+        self.publications: Mapping[str, PublicationArtifact] = (
+            {publication.project_name: publication for publication in publications}
+            if publications
```

Comment:
Setting a default parameter value as a mutable object can lead to surprising behaviour in python, and its [recommended in the docs](https://docs.python.org/3/reference/compound_stmts.html#function-definitions) to use `None` as a default instead: 

> Default parameter values are evaluated from left to right when the function definition is executed. This means that the expression is evaluated once, when the function is defined, and that the same “pre-computed” value is used for each call. This is especially important to understand when a default parameter value is a mutable object, such as a list or a dictionary: if the function modifies the object (e.g. by appending an item to a list), the default parameter value is in effect modified. This is generally not what was intended. A way around this is to use None as the default, and explicitly test for it in the body of the function, e.g.:

```python
def whats_on_the_telly(penguin=None):
    if penguin is None:
        penguin = []
    penguin.append("property of the zoo")
    return penguin
```

We could have callers of the ManifestLoader constructor pass in an empty list instead of `None` but I prefer to have the input sanitization in a single place to make things less error prone / sprawled out.